Added ros interface to blackboard implementation

sasjonge opened the pull request.

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    std::string next_solution() {
+        std::string solution = solutions_.front()->c_str();
```
remove *c_str()* it returns a `char*` and through assignment you convert it to *std::string* again, so the call is effectively not needed

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+        else {
+            std::stringstream ss;
+            ss << *solution->substitution() << ".";
+            solutions_.push_back(std::make_shared<std::string>(ss.str()));
```
are you sure that the format is as expected? In rosprolog, the JSON encoding of the solution string was done in Prolog code, here: https://github.com/knowrob/rosprolog/blob/master/src/query.pl#L36

sasjonge replied:
I rewrote substitutionToJSONString to be used inteast

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+        runQuery(currentQuery_);
+	}
+
+    std::string substitutionToString(const QueryResultPtr &sharedPtr) {
```
maybe it's better to replicate the JSON formatting code here

sasjonge commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+bool finish(json_prolog_msgs::PrologFinish::Request &req,
+            json_prolog_msgs::PrologFinish::Response &res)
+{
+    if (req.id == "*"){
```
It is a string

sasjonge commented:
@daniel86 I think I now incorporated every point of your comments. 

I rewrote a substitution to string function, that works with the rosprolog_commandline.

Additionally I added the tf code. The code compiles now, but I still have to test it.

sasjonge commented:
Something weird happened with the diff in this PR for PrologReasoner.h. I simply added an additional version of the eval function.

daniel86 commented on include/knowrob/ros/tf/logger.h:
```diff
@@ -29,6 +29,9 @@ class TFLogger
 	void set_db_name(const std::string &db_name)
 	{ db_name_ = db_name; }
 
+    void set_db_uri(std::string db_uri)
```
const std::string &db_uri

daniel86 commented on include/knowrob/ros/tf/republisher.h:
```diff
@@ -36,6 +36,9 @@ class TFRepublisher
 	void set_db_name(std::string db_name)
 	{ db_name_ = db_name; }
 
+    void set_db_uri(std::string db_uri)
```
const&

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+using namespace knowrob;
+namespace po = boost::program_options;
+
+class HybridQAResultHandler : public QueryResultHandler {
```
probably should be named differently, *ROSQueryResultHandler* ?

sasjonge replied:
I'm also not completly happy with the name, but i think I want to use the class also for an REST interface later, so ROS doens't fit perfectly. I was thinking about IncrementalQueryResultHandler, but that could create confusion, and make people think the generation of the Results is incremental and not just the next_solution call

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+        incrementalMode = incremental;
+        errorMessage = "";
+        currentQuery_ = queryString;
+        runQuery(currentQuery_);
```
it seems strange that the "result handler" itself is issuing a query. it was supposed to only receive a solution. So the class name HybridQAResultHandler seems not appropriate.

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+        ss << "{\n";
+        for(const auto &pair : *omega) {
+            if(i++ > 0) ss << ",\n";
+            ss << '\"' << pair.first.name() << "\":" << '\"' << (*pair.second) << "\"";
```
sometimes pair.second might be quoted with single quotes! probably you need to remove them in case

sasjonge replied:
I added a filtering for sourrounding single quotes.

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+            solutions_.push_back(std::make_shared<std::string>("True."));
+        }
+        else {
+            std::stringstream ss;
```
stringstream is not needed, just pass the string as argument

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+        else {
+            std::stringstream ss;
+            ss << substitutionToJSONString(solution->substitution());
+            solutions_.push_back(std::make_shared<std::string>(ss.str()));
```
probably unneeded to use smart pointer here

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+            }
+        }
+        catch (std::exception& e) {
+            std::cout << e.what() << std::endl;
```
you should use logging macros, e.g. `KB_ERROR`

daniel86 commented on src/ROSInterface.cpp:
```diff
@@ -0,0 +1,348 @@
+{
+    if (exists(req.id)) {
+        std::stringstream ss;
+        ss << "Another query is already being processed with id " << req.id << ".";
```
the point you can put in single quotes

daniel86 commented on src/ros/tf/logger.cpp:
```diff
@@ -25,13 +25,14 @@ TFLogger::~TFLogger()
 		ROS_WARN("[TFLogger] insert failed: %s.", err.message);
 	}
-	delete collection;
+	collection_.reset();
```
what is the reset doing? is it needed? also for changes below

daniel86 replied:
reset is not needed, just remove it everywhere. smart pointer deletes itself, no reset needed.

daniel86 commented on src/ros/tf/republisher.cpp:
```diff
@@ -35,7 +35,7 @@ TFRepublisher::~TFRepublisher()
 	tick_thread_.join();
 	if(collection_) {
-		delete collection_;
+		collection_.reset();
```
reset is not needed, just remove it everywhere. smart pointer deletes itself, no reset needed.

daniel86 commented on src/ros/tf/republisher.cpp:
```diff
@@ -137,11 +137,12 @@ void TFRepublisher::create_cursor(double start_time)
-	collection_ = MongoInterface::get_collection(
-		db_name_.c_str(),db_collection_.c_str());
+    if(collection_) {
+        collection_.reset();
```
reset is not needed, just remove it everywhere. smart pointer deletes itself, no reset needed.

daniel86 commented on src/ros/tf/republisher.cpp:
```diff
@@ -203,8 +204,8 @@ void TFRepublisher::set_initial_poses(double unix_time)
-	if(collection) {
-		delete collection;
+	if(collection_) {
+		collection_.reset();
```
reset is not needed, just remove it everywhere. smart pointer deletes itself, no reset needed.

sasjonge commented:
@daniel86 I incorporated all your comments and splitted my ROSInterface file so that the ResultHandler is in it's own file (for later use with a REST interface). I now named the ResultHandler IterativeQAResultHandler.
